[sopa301] iP

sopa301 opened the pull request:
# My Commit Straight From the Trash Can

It's figurative trash so don't look at it. Now that's out of the way, the rest of the text is to meet the requirements.

Here's a bullet list:
- One
- Two
- Three

Here's a numbered list:
1. One
2. Two
3. Three

Here's a `Java` code block:
```Java
function void foo() {
    System.out.println("whoops.");
}
```

Here's a task list:
- [X] Enroll in CS.
- [ ] Graduate from CS.
- [ ] Get a job.

Here's an emoji: 🤡 

Here's a quote:
> “Why waste time say lot word when few word do trick?”  -Kevin Malone

Here's a [hyperlink](https://www.youtube.com/watch?v=dQw4w9WgXcQ).

Here's some `inline code`.

Here's some **bold**, *italic* and ~~strikethrough~~ text.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

sopa301 committed:
Changed name of bot from Duke to Bot.

sopa301 committed:
Changed bot personality to trash gremlin Caelus.
Added echo functionality.

sopa301 committed:
Added list capability

sopa301 committed:
Added task marking and unmarking capability

sopa301 committed:
Added helper methods for Task class

sopa301 committed:
Added support for multiple task types

sopa301 committed:
Added test files and re-added echo function.

sopa301 committed:
Added exception handling

sopa301 committed:
Added delete functionality

sopa301 committed:
Minor optimisations

sopa301 committed:
Added documentation, fixed minor bugs

sopa301 committed:
Added data file reading capability

sopa301 committed:
Routed data folder creation to correct level

sopa301 committed:
Merge branch 'branch-Level-7'

sopa301 committed:
Added save functionality

sopa301 committed:
Merge branch 'branch-Level-7'

sopa301 committed:
Added some date functionality

sopa301 committed:
Moved makeTask for classes to the respective classes and added some docs

sopa301 committed:
Merge branch 'branch-Level-8'

sopa301 committed:
Fix style for Task class

Certain excerpts are not conforming to the required style.

Let's change it so that it follows the style guide.

sopa301 committed:
Add TaskList class

We are required to add this abstraction.

Let's add the TaskList class.

sopa301 committed:
Add Storage class

We are required to add this abstraction.

Let's add the Storage class.

sopa301 committed:
Add Ui, Parser and Command classes

As part of the requirements, we have to do this.

Let's add the required classes, the optional Command class,
more linking code and link it all together.

sopa301 committed:
Fix bugs and edit test script

Our code needs to be bug free.

Let's fix the bugs and the test script.

sopa301 committed:
Fix more bugs

Let's fix more bugs.

sopa301 committed:
Merge branch 'branch-A-MoreOOP'

sopa301 committed:
Format code into packages

As per requirement.

Let's put the code in packages.

sopa301 committed:
Merge branch 'branch-A-Packages'

sopa301 committed:
Merge branch 'add-gradle-support'

sopa301 committed:
Add gradle support

As per the requirements.

Let's add gradle and stylecheck.

sopa301 committed:
Add JUnit tests

As per requirements.

Let's add some tests for Task and TaskList.

sopa301 committed:
Merge branch 'branch-A-JUnit'

sopa301 committed:
Add JAR packaging capability

As per requirements, let's make our bot packagable using gradle.

sopa301 committed:
Merge branch 'branch-A-Jar'

sopa301 committed:
Fix coding standard for test files

As per requirements, let's fix the coding standard for TaskListTest
and TaskTest.

sopa301 committed:
Add find command

As per requirements, let's add the find command.

sopa301 committed:
Merge branch 'branch-A-CodingStandard'

sopa301 committed:
Merge branch 'branch-Level-9'

sopa301 committed:
Add code to close scanner

It's nice to close scanners after use. Let's close the scanner in Ui at
the end of its usage.

sopa301 committed:
Fix minor bugs

Let's remove the final keyword from all not-properly-named constants
(thereby making them non-constants) and add code in parser for when
people key in commands without anything else.

sopa301 committed:
Add GUI for the bot

As per requirements, let's add a GUI for our bot.

sopa301 committed:
Fix style

Let's just fix the style for the project because why not?

sopa301 committed:
Move Main out of gui package

Let's move the Main class into the main package because it's the main
class.

sopa301 committed:
Rename package statements

Let's rename the package statements for the classes affected by the
moving of Main so they don't break.

sopa301 committed:
Merge branch 'branch-Level-10'

sopa301 committed:
Fix minor bug

Let's allow the find function to be case-insensitive.

ruiyangzh commented on src/main/java/bot/Bot.java:
```diff
@@ -0,0 +1,84 @@
+     *
+     * @return True if the bot is active, else false.
+     */
+    public boolean isExit() {
```
"Names representing methods must be verbs and written in camelCase."
 Perhaps a method name such as getExitStatus?

ruiyangzh commented on src/main/java/bot/utils/Command.java:
```diff
@@ -0,0 +1,382 @@
+     *
+     * @return True if the bot should exit, false otherwise.
+     */
+    public abstract boolean isExit();
```
As noted.

ruiyangzh commented on src/main/java/bot/utils/TaskList.java:
```diff
@@ -0,0 +1,129 @@
+     *
+     * @return Iterator of Task objects.
+     */
+    public Iterator<Task> iterator() {
```
Perhaps getIterator?

ruiyangzh reviewed (commented):
Hi, not much to comment on, well done generally. (Outside of the scope of this code review, but maybe would be more compartmentalized if you split the child Command classes into their own class files?)

sopa301 committed:
Add assertions to code

Currently, the code has no assertions and the lack of hard-to-find bugs
is purely contingent on good usage of the code, which is hard to achieve
as the codebase grows.

Adding assertions ensures that functions are being used in the way they
are intended.

Let's add assertions to our code. Also because we are required to do so.

sopa301 committed:
Refactor Command and Task classes into packages

Currently, the Command and Tasks classes are written in one file each
with their subclasses.

This is bad because big files are hard to maintain.

Breaking the files up into smaller files will make it easier to
maintain.

Let's break up the Command and Task class files into smaller components
and put them into packages.

We use packages so because of the natural grouping of tasks and
commands.

sopa301 committed:
Improve coding standard and refactor parse() method

Currently, the code has poor coding standard.

Improving the coding standard will make it easier to find bugs.

Let's:
- improve the coding standard as instructed.
- fix the bugs that become apparent during refactoring.

sopa301 committed:
Merge pull request #1 from sopa301/branch-A-CodeQuality

Improve code quality

sopa301 committed:
Merge branch 'master' into branch-A-Assertions

sopa301 committed:
Resolve merge conflicts

Let's resolve merge conflicts so we can merge this commit into master.

sopa301 committed:
Merge pull request #2 from sopa301/branch-A-Assertions

Add assertions to code

sopa301 committed:
Replace old test code and do styling

Let's do a style check so our code adheres to coding standards.

sopa301 committed:
Remove old test code and text-based test from gradle.yaml

Let's remove any text-based testing because we don't need it.

sopa301 committed:
Merge branch 'branch-A-CI'

sopa301 committed:
Make task list contain unique entries

Currently, the task list accepts duplicate tasks.

This is undesirable because it adds no additional information and
clutters the list.

Enforcing uniqueness in the task list will allow us to reduce redundant
information in the list.

Let's make TaskList contain only unique tasks.

sopa301 committed:
Merge pull request #3 from sopa301/branch-C-DetectDuplicates

Make task list contain unique entries

sopa301 committed:
Fix comments to adhere to coding standard

sopa301 committed:
Update README and add UI image

sopa301 committed:
Edit README

sopa301 committed:
Merge branch 'branch-A-UserGuide'

sopa301 committed:
Update main README

sopa301 committed:
Fix cross platform issue

sopa301 committed:
Extend dialog box
